File: ios/support/SingleOwnerResourceSlot.swift
import Foundation

final class SingleOwnerResourceSlot<Resource: AnyObject> {
  private let lock = NSLock()
  private let release: (Resource) -> Void
  private var resource: Resource?

  init(release: @escaping (Resource) -> Void) {
    self.release = release
  }

  var current: Resource? {
    lock.lock()
    defer { lock.unlock() }
    return resource
  }

  @discardableResult
  func replace(_ next: Resource) -> Resource? {
    let previous = swap(next)
    if let previous {
      release(previous)
    }
    return previous
  }

  @discardableResult
  func clear() -> Resource? {
    let previous = take()
    if let previous {
      release(previous)
    }
    return previous
  }

  @discardableResult
  func swap(_ next: Resource) -> Resource? {
    let previous: Resource?
    lock.lock()
    previous = resource
    resource = next
    lock.unlock()
    return previous
  }

  @discardableResult
  func take() -> Resource? {
    let previous: Resource?
    lock.lock()
    previous = resource
    resource = nil
    lock.unlock()
    return previous
  }

  func releaseResource(_ previous: Resource) {
    release(previous)
  }
}

File: android/src/main/java/com/nitroplay/support/SingleOwnerResourceSlot.kt
package com.nitroplay.video.support

internal class SingleOwnerResourceSlot<T>(
  private val release: (T) -> Unit
) {
  private val lock = Any()
  private var resource: T? = null

  val current: T?
    get() = synchronized(lock) { resource }

  fun replace(next: T): T? {
    val previous = swap(next)
    previous?.let(release)
    return previous
  }

  fun clear(): T? {
    val previous = take()
    previous?.let(release)
    return previous
  }

  fun swap(next: T): T? {
    return synchronized(lock) {
      val previous = resource
      resource = next
      previous
    }
  }

  fun take(): T? {
    return synchronized(lock) {
      val previous = resource
      resource = null
      previous
    }
  }

  fun releaseResource(previous: T) {
    release(previous)
  }
}
